File: golang/1656_OrderedStream/OrderedStream.go
// Package main ...
package main

import (
	"reflect"
)

type OrderedStream struct {
	stream []string
	ptr    int
}

func Constructor(n int) OrderedStream {
	return OrderedStream{make([]string, n+1), 1}
}

func (s *OrderedStream) Insert(idKey int, value string) []string {
	s.stream[idKey] = value
	start := s.ptr
	for s.ptr < len(s.stream) && s.stream[s.ptr] != "" {
		s.ptr++
	}
	return s.stream[start:s.ptr]
}

func main() {
	assert := func(a, b []string) {
		if !reflect.DeepEqual(a, b) {
			panic("Not Passed")
		}
	}
	o := Constructor(5)
	assert(o.Insert(3, "ccccc"), []string{})
	assert(o.Insert(1, "aaaaa"), []string{"aaaaa"})
	assert(o.Insert(2, "bbbbb"), []string{"bbbbb", "ccccc"})
	assert(o.Insert(5, "eeeee"), []string{})
	assert(o.Insert(4, "ddddd"), []string{"ddddd", "eeeee"})
}

File: cpp/1656_OrderedStream/OrderedStream.cpp
#include <cassert>
#include <string>
#include <vector>

using namespace std;

class OrderedStream {
 public:
  OrderedStream(int n) {
    stream.resize(n + 1);
    ptr = 1;
  }

  vector<string> insert(int idKey, string value) {
    stream[idKey] = value;
    vector<string> res;
    while (ptr < stream.size() && !stream[ptr].empty()) {
      res.push_back(stream[ptr]);
      ++ptr;
    }
    return res;
  }

 private:
  vector<string> stream;
  int ptr;
};

int main() {
  OrderedStream o{5};
  assert(o.insert(3, "ccccc") == vector<string>{});
  assert(o.insert(1, "aaaaa") == vector<string>{"aaaaa"});
  assert((o.insert(2, "bbbbb") == vector<string>{"bbbbb", "ccccc"}));
  assert((o.insert(5, "eeeee") == vector<string>{}));
  assert((o.insert(4, "ddddd") == vector<string>{"ddddd", "eeeee"}));
}
